SSR-2973: Create user guide of Tabs (#1281)
kintone-labs/kintone-ui-component: docs/document/docs/components/desktop/tabs.md

---
id: tabs
title: Tabs
sidebar_label: Tabs
---

## Overview

A multiple tabs component that can switch displaying contents with buttons.

<div class="sample-container" id="tabs">
  <div id="sample-container__components"></div>
</div>
<script src="/js/samples/desktop/tabs.js"></script>

---

## Specification

### Property
Here is a list of properties that can be used for modifying the component:

> 

Reviewer: I think this order will be better:
1. className
2. id
3. value
4. borderVisible
5. visible
6. items
7. Item.content
8. Item.label
9. Item.value
10. Item.disabled
11. Item.visible

| Name   | Type | Default | Description | Remark |
| :--- | :--- | :--- | :--- | :--- |
| className | string | ""  | Component class name | |
| id | string | ""  | Component id name | |
| value | string | ""  | Component id name | The first tab will be displayed if the `value` is unspecified or there is no matching value in items. |
| borderVisible | boolean | true  | Show/Hide the border surrounding the content | |
| visible | boolean | true | Show/Hide the component | |
| items | Array\<Item> | [] | List of items for tab will be displayed | If the items is not an array, the error message 'items' property is not array. will throw in window console. |
| Item.label | string | "" | Tab name | |

> | Item.value | string | "" | Tab key | `Item.value` is unique and required.<br><li>If `Item.value` is duplicated, the error message 'value' property is not unique in items. will throw in window console.<li>If `Item.value` is NOT set, the error message 'value' property is not specified in items. will throw in window console. |

Reviewer: There is no have end close tag for <li>, please add `</li>`

Reviewer: Please update remark for `Item.value` to:
Will result an error if the `Item.value` is duplicated in columns or not specified.

I refer `Column.field` from Table component: Will result an error if the field is duplicated in columns or not specified

| Item.content | string/HTMLElement | "" | Tab content | |
| Item.visible | boolean | true | Show/Hide the tab | |
| Item.disabled | boolean | true | Enable/Disable the tab | |

### Event

Here is a list of events that can be specified:

| Name | Type | Description | Remark |
| :--- | :--- | :--- | :--- |

> | change | function | Event handler of when selected tab is changed |  It will pass the event object as the argument<br><br>You can receive the following values in event.detail<br><li>event.detail.oldValue : “value” before the change<li>event.detail.value : “value” after the change |

Reviewer: There is no have end close tag for <li>, please add `</li>`

### Constructor

Tabs(options)<br>
Here is a list of available constructors:

#### Parameter

| Name | Type | Default | Description | Remark |
| :--- | :--- | :--- | :--- | :--- |
| options | object | {} | Object that includes component properties | |

---
## Sample Code

> Please check the [package installation](../../getting-started/quick-start.md#installation) method first.

Here is a sample code when all parameters are specified:

> ```javascript

Reviewer: please add `const Kuc = Kucs['1.x.x'];` on top

| Item.visible | boolean | true | Show/Hide the tab | |

### Event

Here is a list of events that can be specified:

| Name | Type | Description | Remark |
| :--- | :--- | :--- | :--- |
| change | function | Event handler of when selected tab is changed |  It will pass the event object as the argument<br><br>You can receive the following values in event.detail<br><li>event.detail.oldValue : “value” before the change</li><li>event.detail.value : “value” after the change</li> |

### Constructor

Tabs(options)<br>
Here is a list of available constructors:

#### Parameter

| Name | Type | Default | Description | Remark |
| :--- | :--- | :--- | :--- | :--- |
| options | object | {} | Object that includes component properties | |

---
## Sample Code

> Please check the [package installation](../../getting-started/quick-start.md#installation) method first.

Here is a sample code when all parameters are specified:

```javascript
const Kuc = Kucs['1.x.x'];
const space = kintone.app.record.getSpaceElement('space');
const tabs = new Kuc.Tabs({
  value:  'tab-a',
  className: 'options-class',
  id: 'options-id',
  visible: true,
  borderVisible: true,

>   items: [

Reviewer: I think we should update Js Sample Code following the content you implemented at desktop/tabs.js
It will be easier for users to visualize the use if it is consistent between the example and the preview

| Name   | Type | Default | Description | Remark |
| :--- | :--- | :--- | :--- | :--- |
| className | string | ""  | Component class name | |
| id | string | ""  | Component id name | |
| value | string | ""  | Component id name | The first tab will be displayed if the `value` is unspecified or there is no matching value in items. |
| borderVisible | boolean | true  | Show/Hide the border surrounding the content | |
| visible | boolean | true | Show/Hide the component | |
| items | Array\<Item> | [] | List of items for tab will be displayed | If the items is not an array, the error message 'items' property is not array. will throw in window console. |
| Item.content | string/HTMLElement | "" | Tab content | |
| Item.label | string | "" | Tab name | |
| Item.value | string | "" | Tab key | `Item.value` is unique and required.<br>Will result an error if the `Item.value` is duplicated in `items` or not specified |
| Item.disabled | boolean | false | Enable/Disable the tab | |
| Item.visible | boolean | true | Show/Hide the tab | |

### Event

Here is a list of events that can be specified:

| Name | Type | Description | Remark |
| :--- | :--- | :--- | :--- |

> | change | function | Event handler of when selected tab is changed |  It will pass the event object as the argument<br><br>You can receive the following values in event.detail<br><li>event.detail.oldValue : “value” before the change</li><li>event.detail.value : “value” after the change</li> |

Reviewer 2: ```suggestion
| change | function | Event handler of when selected tab is changed |  It will pass the event object as the argument<br><br>You can receive the following values in event.detail<br><li>event.detail.oldValue : "value" before the change</li><li>event.detail.value : "value" after the change</li> |
```

### Constructor

Tabs(options)<br>
Here is a list of available constructors:

#### Parameter

| Name | Type | Default | Description | Remark |
| :--- | :--- | :--- | :--- | :--- |
| options | object | {} | Object that includes component properties | |

---
## Sample Code

> Please check the [package installation](../../getting-started/quick-start.md#installation) method first.

Here is a sample code when all parameters are specified:

> ```javascript

Reviewer 2: - Please write all in English.
- Please use a meaningful name for variables and tab name/value, not just a number.
- Please change the sample code scenario for example below.
  - A: TextArea with placeholder
  - B: Dropdown
  - C: just string pattern

> const Kuc = Kucs['1.x.x'];

Reviewer 2: ```suggestion
const Kuc = Kucs['1.x.x'];

```

const space = kintone.app.record.getSpaceElement('space');

const firstContent = document.createElement('div');
const textArea = new Kuc.TextArea({
  label: 'フルーツ',
  requiredIcon: true,
  value: 'Apple',
  error: 'エラーです',
  visible: true,
  disabled: false,
  placeholder: '',
});
firstContent.appendChild(textArea);

const secondContent = document.createElement('div');
const dialog = new Kuc.Dialog({
  title: 'Title',
  content: 'Content with Icon',
  footer: 'Footer',
  icon: 'success',
});
const button = new Kuc.Button({ text: 'Button' });
button.addEventListener('click', () => {
dialog.open();
});
secondContent.appendChild(button);

const thirdContent = document.createElement('div');
thirdContent.innerText = 'tab3_content';
const tabs = new Kuc.Tabs({
  borderVisible: true,
  className: 'kuc-tabs-class',
  id: 'sample-id',
  items: [
    {
      label: 'Tab1',
      content: firstContent,
      value: 'tab1',
      disabled: false,
    },
    {
      label: 'Tab2',
      content: secondContent,

>       value: 'tab2',

Reviewer 2: Please add disabled option.

    },
    {
      value: 'tab3',
      label: 'Tab3',

>       content: thirdContent,

Reviewer 2: Please add disabled option.

---
id: tabs
title: Tabs
sidebar_label: Tabs
---

## Overview

> A multiple tabs component that can switch displaying contents with buttons.

Reviewer 2: How about this to unify with other component guide?

```suggestion
The Tabs component allows the user to display multiple tabs that can switch displaying contents.
```

| value | string | "" | Key of each tab<br>*`Required` and `Unique` | `value` is unique and required.<br>Will result an error if the `value` is duplicated in `items` or not specified |
| disabled | boolean | false | Enable/Disable the tab | |
| visible | boolean | true | Show/Hide the tab | |

### Event

Here is a list of events that can be specified:

| Name | Type | Description | Remark |
| :--- | :--- | :--- | :--- |
| change | function | Event handler of when selected tab is changed |  It will pass the event object as the argument<br><br>You can receive the following values in event.detail<br><li>event.detail.oldValue : "value" before the change</li><li>event.detail.value : "value" after the change</li> |

### Constructor

Tabs(options)<br>
Here is a list of available constructors:

#### Parameter

| Name | Type | Default | Description | Remark |
| :--- | :--- | :--- | :--- | :--- |
| options | object | {} | Object that includes component properties | |

---
## Sample Code

> Please check the [package installation](../../getting-started/quick-start.md#installation) method first.

Here is a sample code when all parameters are specified:

```javascript

> const Kuc = Kucs['1.x.x'];
> 
> const space = kintone.app.record.getSpaceElement('space');
> 
> const firstContent = document.createElement("div");
>   const textArea = new Kuc.TextArea({
>     label: "Fruit",
>     requiredIcon: true,
>     value: "Apple",
>     error: "Error",
>     visible: true,
>     disabled: false,
>     placeholder: "",
>   });
>   firstContent.appendChild(textArea);
> 
>   const secondContent = document.createElement("div");
>   const timePicker = new Kuc.TimePicker({
>     value: "11:30"
>   });
>   secondContent.appendChild(timePicker);
> 
>   const thirdContent = document.createElement("div");
>   thirdContent.innerText = "tab3_content"
> 
>   const tabs = new Kuc.Tabs({
>   borderVisible: true,
>   className: 'kuc-tabs-class',
>   id: 'sample-id',
>   items: [
>     {
>       label: 'TextArea With Placeholder',
>       content: firstContent,
>       value: 'tab-textarea',
>       disabled: false,
>     },
>     {
>       label: 'TimePicker',
>       content: secondContent,
>       value: 'tab-time-picker',
>       disabled: false,
>     },
>     {
>       value: 'tab-string-pattern',
>       label: 'Just String Pattern',
>       content: thirdContent,
>       disabled: false,
>     },
>   ],
>   value: 'tab-textarea',
>   visible: true,
> });
> space.appendChild(tabs);
> 
> tabs.addEventListener('change', event => {
>   console.log(event);
> });

Reviewer 2: How about improving like this? (I haven't tried behavior, so need checking)
- Fixed indent mistake
- Deleted unnecessary div
- Changed tab label, content

If ok, please update component preview, too.

```suggestion
const Kuc = Kucs['1.x.x'];

const space = kintone.app.record.getSpaceElement('space');

// Create each Tab content
const textArea = new Kuc.TextArea({
  label: "TextArea",
  requiredIcon: true,
  value: "This is sample.",
});

const timePicker = new Kuc.TimePicker({
  label: "Time"
  value: "11:30"
});

const text = "This is sample."

const tabs = new Kuc.Tabs({
  items: [
    {
      label: 'A',
      content: textArea,
      value: 'a',
      disabled: false,
    },
    {
      label: 'B',
      content: timePicker,
      value: 'b',
      disabled: false,
    },
    {
      label: 'C',
      content: text,
      value: 'c',
      disabled: false,
    },
  ],
  value: 'a',
  className: 'options-class',
  id: 'options-id',
  visible: true,
  borderVisible: true,
});
space.appendChild(tabs);

tabs.addEventListener('change', event => {
  console.log(event);
});
```

Reviewer 2: Please copy and paste my translated_doc sample code.
It still has differences.
